File: perf/benchmarks/micro/quicksort/quicksort.cpp
// Hand-written quicksort — micro-benchmark (C++).
//
// Algorithmically identical to the Python and Cangjie versions: middle
// element pivot, Lomuto partition, recurse-smaller-side / iterate-larger.
#include <algorithm>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <vector>

static void quicksort(std::vector<int64_t>& a, int64_t lo, int64_t hi) {
    while (lo < hi) {
        int64_t mid = (lo + hi) >> 1;
        int64_t pivot = a[static_cast<size_t>(mid)];
        std::swap(a[static_cast<size_t>(mid)], a[static_cast<size_t>(hi)]);
        int64_t i = lo - 1;
        for (int64_t j = lo; j < hi; ++j) {
            if (a[static_cast<size_t>(j)] <= pivot) {
                ++i;
                std::swap(a[static_cast<size_t>(i)], a[static_cast<size_t>(j)]);
            }
        }
        ++i;
        std::swap(a[static_cast<size_t>(i)], a[static_cast<size_t>(hi)]);
        if ((i - lo) < (hi - i)) {
            quicksort(a, lo, i - 1);
            lo = i + 1;
        } else {
            quicksort(a, i + 1, hi);
            hi = i - 1;
        }
    }
}

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 1'500'000LL;

    std::vector<int64_t> a(static_cast<size_t>(n));
    int64_t seed = 1234567;
    const int64_t mask = 0x7FFFFFFFLL;
    for (int64_t i = 0; i < n; ++i) {
        seed = (seed * 1103515245LL + 12345LL) & mask;
        a[static_cast<size_t>(i)] = seed;
    }

    auto t0 = std::chrono::steady_clock::now();
    quicksort(a, 0, n - 1);
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    int64_t cs = 0;
    for (int64_t k = 0; k < n; k += 1024) {
        cs ^= a[static_cast<size_t>(k)];
    }
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/micro/quicksort/quicksort.py
# Hand-written quicksort — micro-benchmark.
#
# This is a complement to the `sort` benchmark (which uses the language's
# stdlib sort): here we run an algorithmically identical Lomuto-partition
# quicksort in all three languages so that ALL cross-language work — pivot
# choice, partition step, recursion shape, and array indexing — is the same.
#
# The pivot is the middle element; recursion is on the smaller side first
# (with iterative tail-call on the larger side) so worst-case stack depth is
# O(log N) and the algorithm is deterministic given the input.
#
# Argument: N (default 1_500_000).

import sys
import time


def quicksort(a: list, lo: int, hi: int) -> None:
    while lo < hi:
        # Lomuto partition with middle-element pivot.
        mid = (lo + hi) >> 1
        pivot = a[mid]
        # Move pivot to the end.
        a[mid], a[hi] = a[hi], a[mid]
        i = lo - 1
        for j in range(lo, hi):
            if a[j] <= pivot:
                i += 1
                a[i], a[j] = a[j], a[i]
        i += 1
        a[i], a[hi] = a[hi], a[i]
        # Recurse on the smaller side first; iterate on the other.
        if (i - lo) < (hi - i):
            quicksort(a, lo, i - 1)
            lo = i + 1
        else:
            quicksort(a, i + 1, hi)
            hi = i - 1


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 1_500_000

    # Deterministic LCG (kept outside the timed region).
    a = [0] * n
    seed = 1234567
    mask = 0x7FFFFFFF
    for i in range(n):
        seed = (seed * 1103515245 + 12345) & mask
        a[i] = seed

    sys.setrecursionlimit(max(1000, n))
    t0 = time.perf_counter_ns()
    quicksort(a, 0, n - 1)
    elapsed_ns = time.perf_counter_ns() - t0

    # Checksum: every 1024-th element XORed in.
    cs = 0
    k = 0
    while k < n:
        cs ^= a[k]
        k += 1024
    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
